Correctly support UseWikiProfile

cpeel opened the pull request:
The conversion to PluggableAuth broke support for UseWikiProfile. This adds it back in a way that is compatible with PluggableAuth. This version allows a user to log in with either their phpBB username or, if set and `UseWikiProfile = true`, their WikiProfile username.

I tested:
* `UseWikiProfile = false`
* `UseWikiProfile = true`
  * Logging in with the phpBB username (WikiProfile username is used for account create if set)
    * Using a phpBB username within invalid MediaWiki characters (spaces)
    * Using phpBB usernames with non-Latin-1 characters
  * Logging in with the WikiProfile username
* Using `UseWikiGroup = true`

Edge cases handled:
* More than one user has the same WikiProfile username.
  * If one of the users logs in with their phpBB username this works fine since those are unique
  * If one of the users logs in with the duplicate WikiProfile username, login fails and we write an error to `php_errors`.

cpeel committed:
Correctly support UseWikiProfile

The conversion to PluggableAuth broke support for UseWikiProfile. This
adds it back in a way that is compatible with PluggableAuth.

JWPlatt commented:
@cpeel It looks like I didn't think about duplicate profile names.  Nice work.  I wrote up some conditions to check:

1. pbpBB account profile may be blank to mean the user's phpBB and wiki account login names are identical.
2. phpBB account profile contains valid wiki login name if different from phpBB username.
3. phpBB account profile may be same as the phpBB username and the redundancy is used and tolerated, though unnecessary.
4. phpBB account profile name and entered password is used to create wiki database account username and password at login if the account doesn't already exist.
5. If phpBB profile is blank, wikilogin uses wikilogin name as phpBB username for authentication and pbpBB usergroup membership lookup.
6. If phpBB profile is not blank, wikilogin name is translated via profile to pbpBB username for authentication and pbpBB usergroup membership lookup.
7. If authenticated member is not in the usergroup, and the requirement is enabled, they should not be able to edit the wiki.
8. phpBB username mapped to existing wikiusername should use existing wiki account without causing a new wiki account to be created.
9. wikilogin name not found in non-blank phpBB profile OR not found in phpBB user table is denied wiki login without creating a new account.
10. All wiki accounts should still work, possibly with password recovery, if the profile translation feature is disabled or the entire extension is removed.

cpeel commented:
@Digitalroot - As the lead developer of [Distributed Proofreaders](https://github.com/DistributedProofreaders/dproofreaders) I have an incentive to keep this plugin alive as we use it at pgdp.net. If you're amenable to me assisting with its overall maintenance and up-keep I'd welcome a conversation about how to make that possible. I can be reached via mail using my github username at the domain kence.org.

JWPlatt commented:
@cpeel  7. My memory was that login was allowed but editing was not.  I could be wrong.  They can still use the wiki logged out, so not a huge deal.

10. I had to remove the plugin when it was broken by the wiki upgrade.  I have had no complaints about logins/missing accounts or password recovery.  It's possible some just figured they forgot their password and recovered it via the wiki.  So long as password recovery works, great.

cpeel commented:
> 7. My memory was that login was allowed but editing was not. I could be wrong. They can still use the wiki logged out, so not a huge deal.

The [v3.x code](https://github.com/Digitalroot-Technologies/MediaWiki_PHPBB_Auth/blob/v3.x/Auth_phpBB.php#L426) only prevents login and does not influence authorization and this retains that functionality.

> 10. I had to remove the plugin when it was broken by the wiki upgrade.  I have had no complaints about logins/missing accounts or password recovery.  It's possible some just figured they forgot their password and recovered it via the wiki.  So long as password recovery works, great.

I will do some due-diligence on this but have very high confidence that this will Just Work based on the status of the MediaWiki accounts that are created by PluggableAuth.

cpeel commented:
I did more testing this evening and fixed one minor issue I found. I consider this good to merge now.

cpeel committed:
Update UseWikiProfile documentation

Go into much more detail about how Auth_phpBB maps usernames and
how and when to use the UseWikiProfile option.

cpeel commented:
One last commit in the PR to update the `README.md` on the use of `UseWikiProfile`.
